src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -192,7 +194,146 @@ size_t BestBinaryNumericSplit<FitnessFunction>::CalculateDirection(
     const arma::Col<ElemType>& classProbabilities,
     const AuxiliarySplitInfo<ElemType>& /* aux */)
 {
-  if (point <= classProbabilities[0])
+  if (point <= classProbabilities(0))
+    return 0; // Go left.
+  else
+    return 1; // Go right.
+}
+
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double RandomBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    arma::Col<typename VecType::elem_type>& classProbabilities,
+    AuxiliarySplitInfo<typename VecType::elem_type>& /* aux */)
+{
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  // Forcing a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimum * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  typename VecType::elem_type maxValue = arma::max(data);
+  typename VecType::elem_type minValue = arma::min(data);
+
+  // Sanity check: if the maximum element is the same as the mininimum, we
+  // can't split in this dimension.
+  if (maxValue == minValue)
+    return DBL_MAX;
+
+  /*
+  Just for making review easy, the following bit of code is taken directly from
+  https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution
+  to generate a random number. (To be removed before merge)
+  */
+  // Picking a random pivot to split the dimension.
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> distribution(minValue, maxValue);
+  double randomPivot = distribution(gen);
```

Comment:
Here I am generating a random point between the minimum and maximum value of the feature (as mentioned in the research paper). Thus, the `randomPivot` might or might not be a point belonging to the dataset.

Another way to achieve a somewhat similar thing could be to generate a random index between `0` and `data.n_elem - 1` and then `randomPivot` could be the point belonging to that index in the data.

Which one seems like a better approach here ?

Reply:
Yeah sure. I can use that. Thanks!